pallets/xvm/src/lib.rs

```diff
@@ -0,0 +1,302 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Astar. If not, see <http://www.gnu.org/licenses/>.
+
+//! # XVM pallet
+//!
+//! ## Overview
+//!
+//! ## Interface
+//!
+//! ### Dispatchable Function
+//!
+//!
+//! ### Other
+//!
+//!
+
+use frame_support::{pallet_prelude::*, traits::ConstU32, BoundedVec};
+use pallet_evm::GasWeightMapping;
+use parity_scale_codec::{Decode, Encode};
+use scale_info::TypeInfo;
+use sp_core::U256;
+use sp_runtime::{traits::StaticLookup, RuntimeDebug};
+use sp_std::{marker::PhantomData, prelude::*, result::Result};
+
+use astar_primitives::ethereum_checked::{
+    AccountMapping, CheckedEthereumTransact, CheckedEthereumTx, MAX_ETHEREUM_TX_INPUT_SIZE,
+};
+
+pub use pallet::*;
+
+// TODO: move types and traits to `astar-primitives`.
+
+/// XVM call info on success.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub struct CallInfo {
+    /// Output of the call.
+    pub output: Vec<u8>,
+    /// Actual used weight.
+    pub used_weight: Weight,
+}
+
+/// XVM call error on failure.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub enum CallError {
+    /// The call failed on EVM or WASM execution.
+    ExecutionFailed(Vec<u8>),
+    /// Input is too large.
+    InputTooLarge,
+    /// Target contract address is invalid.
+    InvalidTarget,
+    /// Calling the contracts in the same VM is not allowed.
+    SameVmCallNotAllowed,
+}
+
+/// XVM call error with used weight info.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub struct CallErrorWithWeight {
+    /// Error info.
+    pub error: CallError,
+    /// Actual used weight.
+    pub used_weight: Weight,
+}
+
+/// XVM call result.
+pub type XvmCallResult = Result<CallInfo, CallErrorWithWeight>;
+
+#[derive(PartialEq, Eq, Copy, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub enum Vm {
+    Evm,
+    Wasm,
+}
+
+// TODO: Note caller shouldn't be able to specify `source_vm`.
+/// XVM context.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub struct XvmContext {
+    /// The source VM of the call.
+    pub source_vm: Vm,
+    /// The target VM of the call.
+    pub target_vm: Vm,
+    /// Max weight limit.
+    pub weight_limit: Weight,
+    /// Optional encoded execution environment.
+    pub env: Option<Vec<u8>>,
+}
+
+pub trait XvmCall<AccountId> {
+    /// Call a contract in XVM.
+    ///
+    /// Parameters:
+    /// - `context`: XVM context.
+    /// - `source`: Caller Id.
+    /// - `target`: Target contract address.
+    /// - `input`: call input data.
+    fn xvm_call(
+        context: XvmContext,
+        source: AccountId,
+        target: Vec<u8>,
+        input: Vec<u8>,
```

Comment:
Same for the input, it can be a enum with different tuple variants to accommodate different VMs. 
since `pallet_contracts` also expects storage deposit value, which currently is set to `None`, but will default to a conservative value in the upcoming runtime release, so users should be able to change it too.
```rust
pub enum CallInput {
     Evm {value: u128, input: Vec<u8>}
     Wasm {value: Balance, input: Vec<u8>, storage_deposit: Option<Balance>}
}
```

Reply:
This is a draft for discussion so I didn't make much change with interface details.

I'm open to updating parameter types. Potentially the enum definition will increase the complexity of encoding/decoding between VMs and runtime. I'll try different solutions and see.